### PR TITLE
Feature/add rules 2

### PR DESCRIPTION
**Pull request by dereuromark**

### CODE REVIEW

File: src/Common/Factory/CreateContainOneNewFactoryRule.php [hunk 1]
```diff
@@ -0,0 +1,101 @@
+<?php
+
+namespace ArchitectureSniffer\Common\Factory;
+
+use PHPMD\AbstractNode;
+use PHPMD\Node\MethodNode;
+use PHPMD\Rule\MethodAware;
+
+/**
+ * Every Facade should only retrieve native types and transfer objects
```
Comment (stereomon) [inline src/Common/Factory/CreateContainOneNewFactoryRule.php]: The description is wrong here, please change it.

File: src/Common/Factory/CreateContainOneNewFactoryRule.php [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+     *
+     * @return void
+     */
+    private function applyRule(MethodNode $method)
+    {
+        if (substr($method->getName(), 0, 6) !== 'create') {
+            return;
+        }
+
+        $count = count($method->findChildrenOfType('AllocationExpression'));
```
Comment (stereomon) [inline src/Common/Factory/CreateContainOneNewFactoryRule.php]: PLease extract the following three checks into a method.

File: src/Common/Factory/CreateContainOneNewFactoryRule.php [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+        $firstPrimaryPrefix = $primaryPrefixes[0];
+        return $firstPrimaryPrefix->getChild(0)->getName() === 'parent';
+    }
+
+    /**
+     * @param \PHPMD\Node\MethodNode $method
+     *
+     * @return bool
+     */
+    protected function isIndirectFactoryMethod(MethodNode $method)
```
Comment (stereomon) [inline src/Common/Factory/CreateContainOneNewFactoryRule.php]: What means indirectFactoryMethod? Here would a description helpful.

File: src/SprykerAbstractRule.php [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+ * @license http://www.opensource.org/licenses/bsd-license.php BSD License
+ */
+
+namespace ArchitectureSniffer;
+
+use PHPMD\AbstractRule;
+use PHPMD\Node\MethodNode;
+
+/**
+ * TODO
```
Comment (stereomon) [inline src/SprykerAbstractRule.php]: Remove todo, do we need the file doc block like this here?

File: src/Zed/Business/Facade/ZedBusinessFacadeRule.php [hunk 5]
```diff
@@ -21,36 +20,12 @@ public function apply(AbstractNode $node)
...
             $this->applyNoInstantiationsWithNew($method);
         }
     }
 
     /**
-     * @param \PHPMD\Node\ClassNode $class
-     *
-     * @return void
-     */
-    private function applyStatelessThereAreNoProperties(ClassNode $class)
```
Comment (stereomon) [inline src/Zed/Business/Facade/ZedBusinessFacadeRule.php]: Why is that one removed? Facades should not have properties.

File: src/Zed/Business/Factory/MethodReturnInterfaceRule.php [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+     * @param \PHPMD\Node\MethodNode $node
+     *
+     * @return void
+     */
+    private function applyRule(MethodNode $node)
+    {
+        $comment = $node->getComment();
+        if ($this->hasInvalidReturnType($comment)) {
+            $message = sprintf(
+                'TODO The %s is using a invalid return type "%s" which violates the rule "Should only return interfaces"',
```
Comment (stereomon) [inline src/Zed/Business/Factory/MethodReturnInterfaceRule.php]: Remove todo

File: src/Zed/Common/Factory/StatelessFactoryRule.php [hunk 7]
```diff
@@ -1,52 +0,0 @@
...
-
-use PHPMD\AbstractNode;
-use PHPMD\AbstractRule;
-use PHPMD\Node\ClassNode;
-use PHPMD\Rule\ClassAware;
-
-/**
- * Factories shouldn't contain any property. Properties would make the factory stateful.
- */
-class StatelessFactoryRule extends AbstractRule implements ClassAware
```
Comment (stereomon) [inline src/Zed/Common/Factory/StatelessFactoryRule.php]: Is that check moved to somewhere else? I think the rule is valid and shouldn't be removed.

File: src/Zed/Dependency/Bridge/BridgeConstructorArgumentsRule.php [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+     *
+     * @return void
+     */
+    private function applyRule(MethodNode $method)
+    {
+        if ($method->getName() !== '__construct') {
+            return;
+        }
+
+        $params = $method->getParameters();
```
Comment (stereomon) [inline src/Zed/Dependency/Bridge/BridgeConstructorArgumentsRule.php]: This one should be extracted in a separate rule. It's a violation of SRP. Every Rule should only contain one check.

File: src/Zed/ruleset.xml [hunk 9]
```diff
@@ -12,117 +12,99 @@
...
     <rule
-        name="ArgumentsFacadeRule"
-        message="Zed Business Layer - Facade: {0}"
-        class="ArchitectureSniffer\Zed\Business\Facade\ArgumentsFacadeRule">
+            name="ArgumentsFacadeRule"
+            message="Zed Business Layer - Facade: {0}"
+            class="ArchitectureSniffer\Zed\Business\Facade\ArgumentsFacadeRule">
 
-        <priority>2</priority>
+        <priority>3</priority>
```
Comment (stereomon) [inline src/Zed/ruleset.xml]:
```
We decided to use two priorities:
Prio 1: API-related (e.g. something in a facade) OR runtime-critical (e.g. "No call from Client to Zed")
Prio 2: Everyhing else
```
Should we mention somewhere that we have also 3? Is that for currently not fixable rules?

File: src/Common/Factory/CreateContainOneNewFactoryRule.php (hunk 1 above)
Comment (dereuromark) [inline src/Common/Factory/CreateContainOneNewFactoryRule.php]: :) thx

File: src/Zed/Business/Facade/ZedBusinessFacadeRule.php (hunk 5 above)
Comment (dereuromark) [inline src/Zed/Business/Facade/ZedBusinessFacadeRule.php]: I can revert here.

File: src/Zed/Common/Factory/StatelessFactoryRule.php (hunk 7 above)
Comment (dereuromark) [inline src/Zed/Common/Factory/StatelessFactoryRule.php]: The rule is invalid and utterly wrong, the caching of (immutable) factories and builders is exactly doing the opposite of what the text here claims: It actually prevents statefulness.
For factories the properties must stay possible.

File: src/Zed/ruleset.xml (hunk 9 above)
Comment (dereuromark) [inline src/Zed/ruleset.xml]: Yes: https://github.com/spryker/architecture-sniffer#levels
and also soon in the high level wrapper in Development module.

File: src/Zed/Dependency/Bridge/BridgeConstructorArgumentsRule.php (hunk 8 above)
Comment (dereuromark) [inline src/Zed/Dependency/Bridge/BridgeConstructorArgumentsRule.php]: done

File: src/Common/Factory/CreateContainOneNewFactoryRule.php (hunk 2 above)
Comment (dereuromark) [inline src/Common/Factory/CreateContainOneNewFactoryRule.php]: I dont see the method here tbh.

File: src/Zed/Common/Factory/StatelessFactoryRule.php (hunk 7 above)
Comment (stereomon) [inline src/Zed/Common/Factory/StatelessFactoryRule.php]: if we need some sort of caching we should find a proper solution for the real problem. Please keep this Rule until we can clarify with @FabianWesner

Comment (dereuromark) [inline src/Zed/Common/Factory/StatelessFactoryRule.php]: I already did before he left. On this one.
We can add another more sane (with proper explanation) rule in the future here.

**Pull request merged**